### PR TITLE
ARROW-9548: [Go] Test output files are not removed correctly

### PR DESCRIPTION
**Pull request by fredgan**
@sbinet Please review it

### CODE REVIEW

File: go/arrow/arrio/arrio_test.go [hunk 1]
```diff
@@ -98,15 +98,15 @@ func TestCopy(t *testing.T) {
 							if err != nil {
 								t.Fatal(err)
 							}
-							defer f.Close()
 							defer os.Remove(f.Name())
+							defer f.Close()
```
Comment (sbinet) [inline go/arrow/arrio/arrio_test.go:102]: hum... perhaps we should just create a per-sub-test tmp directory and `defer os.RemoveAll(tmpdir)` instead?
(and loose the `defer os.Remove(file.Name())` calls)

File: go/arrow/ipc/cmd/arrow-json-integration-test/main.go [hunk 2]
```diff
@@ -158,16 +148,6 @@ func cnvToARROW(arrowName, jsonName string, verbose bool) error {
 		return xerrors.Errorf("invalid number of records copied (got=%d, want=%d", got, want)
 	}
 
-	err = ww.Close()
```
Comment (sbinet) [inline go/arrow/ipc/cmd/arrow-json-integration-test/main.go:161]: ditto

File: go/arrow/ipc/cmd/arrow-json-integration-test/main_test.go [hunk 3]
```diff
@@ -39,8 +39,8 @@ func TestIntegration(t *testing.T) {
 			if err != nil {
 				t.Fatal(err)
 			}
+			defer os.Remove(af1.Name())
 			defer af1.Close()
-			defer os.RemoveAll(af1.Name())
```
Comment (sbinet) [inline go/arrow/ipc/cmd/arrow-json-integration-test/main_test.go:43]: same comment than above: perhaps replace with a per-sub-test top-level tmp dir that is being `defer os.RemoveAll(tmpdir)` ?

**Comment by fredgan** [comment]
@sbinet Yeah, I agree. I have changed them to this. Should be OK now.

File: go/arrow/arrio/arrio_test.go [hunk 4]
```diff
@@ -67,6 +67,12 @@ func (k copyKind) check(t *testing.T, f *os.File, mem memory.Allocator, schema *
 func TestCopy(t *testing.T) {
 	type kind int
 
+	tempDir, err := ioutil.TempDir("", "arrow-ipc-")
```
Comment (sbinet) [inline go/arrow/arrio/arrio_test.go:70]:
```suggestion
	tempDir, err := ioutil.TempDir("", "arrow-io-")
```

File: go/arrow/arrio/arrio_test.go [hunk 5]
```diff
@@ -94,19 +100,17 @@ func TestCopy(t *testing.T) {
 							mem := memory.NewCheckedAllocator(memory.NewGoAllocator())
 							defer mem.AssertSize(t, 0)
 
-							f, err := ioutil.TempFile("", "arrow-ipc-")
+							f, err := ioutil.TempFile(tempDir, "arrow-ipc-")
```
Comment (sbinet) [inline go/arrow/arrio/arrio_test.go:103]:
```suggestion
							f, err := ioutil.TempFile(tempDir, "arrow-io-")
```

or just use the test name?

File: go/arrow/arrio/arrio_test.go [hunk 6]
```diff
@@ -94,19 +100,17 @@ func TestCopy(t *testing.T) {
...
 							if err != nil {
 								t.Fatal(err)
 							}
 							defer f.Close()
-							defer os.Remove(f.Name())
 
-							o, err := ioutil.TempFile("", "arrow-ipc-")
+							o, err := ioutil.TempFile(tempDir, "arrow-ipc-")
```
Comment (sbinet) [inline go/arrow/arrio/arrio_test.go:109]: ditto

File: go/arrow/internal/arrjson/arrjson_test.go [hunk 7]
```diff
@@ -36,12 +42,11 @@ func TestReadWrite(t *testing.T) {
 			mem := memory.NewCheckedAllocator(memory.NewGoAllocator())
 			defer mem.AssertSize(t, 0)
 
-			f, err := ioutil.TempFile("", "arrjson-")
+			f, err := ioutil.TempFile(tempDir, "arrjson-")
```
Comment (sbinet) [inline go/arrow/internal/arrjson/arrjson_test.go:45]: or use the test name to derive a file name?

File: go/arrow/ipc/cmd/arrow-cat/main_test.go [hunk 8]
```diff
@@ -174,7 +180,7 @@ record 3...
 			defer mem.AssertSize(t, 0)
 
 			fname := func() string {
-				f, err := ioutil.TempFile("", "go-arrow-")
+				f, err := ioutil.TempFile(tempDir, "go-arrow-")
```
Comment (sbinet) [inline go/arrow/ipc/cmd/arrow-cat/main_test.go:183]: ditto

File: go/arrow/ipc/cmd/arrow-cat/main_test.go [hunk 9]
```diff
@@ -515,7 +526,7 @@ record 3/3...
...
```
Comment (sbinet) [inline go/arrow/ipc/cmd/arrow-cat/main_test.go:529]: ditto.

File: go/arrow/ipc/cmd/arrow-file-to-stream/main_test.go [hunk 10]
```diff
@@ -27,25 +27,30 @@ import (
...
+	}
+	defer os.RemoveAll(tempDir)
+
 	for name, recs := range arrdata.Records {
 		t.Run(name, func(t *testing.T) {
 			mem := memory.NewCheckedAllocator(memory.NewGoAllocator())
 			defer mem.AssertSize(t, 0)
 
-			f, err := ioutil.TempFile("", "arrow-ipc-")
+			f, err := ioutil.TempFile(tempDir, "arrow-ipc-")
```
Comment (sbinet) [inline go/arrow/ipc/cmd/arrow-file-to-stream/main_test.go:41]: ditto, and for all occurrences below.

File: go/arrow/internal/arrjson/arrjson_test.go (hunk 7 above)
Comment (fredgan) [reply go/arrow/internal/arrjson/arrjson_test.go]: @sbinet Well, I prefer the test function name. And maybe add "go-arrow-" prefix will be better.

Comment (sbinet) [reply go/arrow/internal/arrjson/arrjson_test.go]: fair enough.

**Pull request closed without merging**